### DISCUSSION TITLE
How enable CORS in server API?

### DISCUSSION THREAD

**Discussion by ahku** [nuxt/nuxt#17484] [Questions]
I would like to enable CORS like you do it for express (https://expressjs.com/en/resources/middleware/cors.html) but can't figure out the right way to do it. 

My use case is that I'm trying to create some API endpoints that should be accessible to other domains. Preferably I should also be able to control this access 🤠.

Thankful for any help.

**Comment by mihail727** [comment]
the old solution was:
```ts
import { createApp, createRouter, useQuery, useBody } from 'h3';
import cors from 'cors';

const app = createApp();
const router = createRouter();

app.use(cors());
app.use(router);

router.get('/users', async () => await userService.getUsers());

export default app;
```

but how to work with it now .... also interesting

**Comment by kakajansh** [comment]
Should be possible with [nitro](https://github.com/unjs/nitro) plugins

1. Add [cors](https://www.npmjs.com/package/cors) dependency to your project
2. Create custom plugin at `/server/plugins/cors.ts`

```ts
import cors from 'cors'

export default defineNitroPlugin(({ h3App }) => {
  h3App.use(cors)
})
```

3. Add plugin to your `nuxt.config.ts`
```ts
export default defineNuxtConfig({
  nitro: {
    plugins: [
      '~/server/plugins/cors'
    ]
  }
})
```

**Reply by mihail727** [reply]
what a about multer? do you know how to use it with routes?

**Reply by kakajansh** [reply]
Have not used `multer`, but below code uses its alternative `formidable`. This route accepts single image and sends it to 3rd party api. Functionality not tested, but it should give you some idea.
 
```ts
import fs from 'fs'
import axios from 'axios'
import { Readable } from 'stream'
import formidable from 'formidable'
import { FormData } from 'formdata-node'
import { FormDataEncoder } from 'form-data-encoder'
import { fileFromPath } from 'formdata-node/file-from-path'

export default defineEventHandler(async (event) => {
  try {
    const form = formidable({
      keepExtensions: true,
      multiples: false,
      maxFileSize: 2 * 1024 * 1024, // 2 MB in bytes
    })

    // handle form parse. Returns image file
    const parseForm = new Promise(async (resolve, reject) => {
      form.parse(event.req, (err, fields, files) => {
        if (err) {
          reject(err)
          return;
        }

        if (!files.imageFile) {
          reject(new Error('No image file'))
          return;
        }

        resolve(files.imageFile)
      });
    })

    const imageFile = await parseForm
    const formData = new FormData()
    formData.append('imageFile', await fileFromPath(imageFile?.filepath))
    const encoder = new FormDataEncoder(formData)

    // send request to vin api
    const { data } = await axios.request({
      method: 'post',
      url: 'API_ENDPOINT',
      data: Readable.from(encoder),
      headers: {
        'content-type': encoder.contentType,
        'content-length': encoder.contentLength,
        'x-custom-header': 'header-value'
      }
    })

    // remove stored file
    fs.unlinkSync(imageFile?.filepath)

    return {
      success: true,
      data,
    }
  } catch (error) {
    event.res.statusCode = error.httpCode ?? 500
    event.res.statusMessage = error.message ?? 'Server Error'
    return {
      success: false,
      message: error.message ?? 'Server Error'
    }
  }
})

```

**Reply by ahku** [reply]
@kakajansh Unfortunately I still could not get this to work, but it feels close! The way I test this is to try to access the endpoint from another localhost address with a different port. And this endpoint access gets blocked by CORS.

**Reply by mihail727** [reply]
> 

nah..... same problem .... seems like limit size
its working only with small files.... with large i have an error: 
```
 ERROR  [h3] write EPIPE                                                                                                                                     17:21:55

  at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16
```

**Reply by ahku** [reply]
@kakajansh 
I've now confirmed that the following works with an express server (notice the origin value set to localhost):

```js
app.use(
  cors({
    origin: 'http://localhost:3100',
    credentials: true,
  })
)
app.use(function(req, res, next) {
  res.header('Access-Control-Allow-Origin', 'http://localhost:3100')
  res.header('Access-Control-Allow-Credentials', true)
  res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE,OPTIONS')
  res.header(
    'Access-Control-Allow-Headers',
    'Origin,X-Requested-With,Content-Type,Accept,content-type,application/json'
  )
  next()
})
```

But the Nitro plugin version does not, and the client compains that the origin is set to '*' even though I set it to localhost:

```js
import cors from 'cors'

export default defineNitroPlugin(({ h3App }) => {
  h3App.use(
    cors({
      origin: 'http://localhost:3100',
      credentials: true,
    })
  )
  h3App.use(function (req, res, next) {
    res.header('Access-Control-Allow-Origin', 'http://localhost:3100')
    res.header('Access-Control-Allow-Credentials', true)
    res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE,OPTIONS')
    res.header(
      'Access-Control-Allow-Headers',
      'Origin,X-Requested-With,Content-Type,Accept,content-type,application/json'
    )
    next()
  })
})
```

I'm unfortunately not familiar enough with how Nitro plugins works and in which order things are executed, but perhaps you can spot the difference here?

**Reply by ahku** [reply]
> Should be possible with [nitro](https://github.com/unjs/nitro) plugins
> 
> 1. Add [cors](https://www.npmjs.com/package/cors) dependency to your project
> 2. Create custom plugin at `/server/plugins/cors.ts`
> 
> ```ts
> import cors from 'cors'
> 
> export default defineNitroPlugin(({ h3App }) => {
>   h3App.use(cors)
> })
> ```
> 
> 3. Add plugin to your `nuxt.config.ts`
> 
> ```ts
> export default defineNuxtConfig({
>   nitro: {
>     plugins: [
>       '~/server/plugins/cors'
>     ]
>   }
> })
> ```

@kakajansh I finally got this to work between two localhost apps, using the `credentials: 'omit'` option in `fetch()`! However, when deploying to Netlify the same solution doesn't seem to work. At this point maybe it's something to do with the Nitro -> Netlify adapter.

**Reply by ahku** [reply]
False alarm, by setting `credentials: 'omit'` I got it to work with fetch() on localhost, but it was in fact not reading my header origin set via the Nitro plugin at all.

**Comment by wobsoriano** [comment]
How about just exporting the cors middleware via a server middleware?

```ts
// ~/server/middleware/cors.ts

import cors from 'cors'

export default cors({ origin: 'http://localhost:3000' })
```
